importlib_resources/_compat.py

```diff
@@ -76,7 +80,8 @@ def _file_reader(spec):
             or
             # local FileReader
             _file_reader(self.spec)
-            or _adapters.DegenerateFiles(self.spec)
+            # fallback - adapt the spec ResourceReader to TraversableReader
+            or _adapters.CompatibilityFiles(self.spec)
```

Comment:
The `_compat` module is not part of the CPython implementation, so this compatibility shim will not have any effect once this is ported to CPython. As a result, a third-party loader providing `get_resource_reader` without `.files` will not stop working for the legacy API. To support this use-case, I suspect we'll want to move the `_native_reader() or CompatibilityFiles()` logic to `_common.from_package`. Maybe also `_file_reader()` (I'm trying to imagine what loader might have modules on the file system but not be using the built-in FileReader), but I'd say let's start with just `_native_reader` and `CompatibiltyFiles`. Consider implementing this compatibility shim as a `CompatibilityFiles.try_native` classmethod.

Let me know if you'd like me to take a stab at the implementation or clarify.

Also, I note that this need is actually somewhat orthogonal to this change. The implementation you have here is perfectly suitable for `importlib_resources` and just needs to be extended for `importlib.resources`, which I might decide to defer in order to get some feedback on the approach as written.

Reply:
I can have a look at it later, but I think that could be done in a separate PR. We just have to coordinate if you want to sync with CPython soon.

Reply:
> this compatibility shim will not have any effect once this is ported to CPython

On further consideration, I believe the tests already address this concern and here's how:

- [create_package](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/util.py#L63-L67) in util creates a Loader/Reader that [does not implement `.files()`](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/util.py#L17-L51).
- For that loader/reader [tests assert](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/tests/test_compatibilty_files.py#L87-L89) that a `CompatibilityFiles` is loaded by `wrap_spec()` (as found in _adapters).
- [files() calls wrap_spec](https://github.com/python/importlib_resources/blob/54a61c457df31c4e0e3ac38aa6ebc6fe1acb671d/importlib_resources/_common.py#L18-L23), which [on CPython is `_adapters.wrap_spec`](https://github.com/python/importlib_resources/blob/d89e03e3800a4baf50e3f3820754046971d04ae7/Lib/importlib/_common.py#L12).

> To support this use-case, I suspect we'll want to move the `_native_reader() or CompatibilityFiles()` logic to `_common.from_package`

- The `_native_reader() or CompatibilityFiles()` [logic is present](https://github.com/python/importlib_resources/blob/d89e03e3800a4baf50e3f3820754046971d04ae7/Lib/importlib/_adapters.py#L63-L68), similar to how I was imagining:

> Consider implementing this compatibility shim as a `CompatibilityFiles.try_native` classmethod.

At this point, I no longer have any concerns about the compatibility shim not being active in the stdlib.

Reply:
I was just about to comment on the bpo, yes `CompatibilityFiles._native_reader` already provides the fallback on readers with `files()`.

My concern here would maybe be the stdlib using non-`TraversableResources` readers where a more specialized reader is available (eg. `SourceFileLoader`), but I think `CompatibilityFiles` provides a perfectly adequate shim for those. If there are any issues because of this, the best solution would be to replace those instances with `TraversableResources` readers.